File: leetcode/subtree-of-another-tree/subtree-of-another-tree-naive-iter.rb
# LeetCode #572 - Subtree of Another Tree
# https://leetcode.com/problems/subtree-of-another-tree/
# Naive solution, quadratic O(MN) runtime, implemented iteratively.

# Definition for a binary tree node.
# class TreeNode
#     attr_accessor :val, :left, :right
#     def initialize(val = 0, left = nil, right = nil)
#         @val = val
#         @left = left
#         @right = right
#     end
# end
# @param {TreeNode} root
# @param {TreeNode} sub_root
# @return {Boolean}
def is_subtree(root, sub_root)
  return true unless sub_root

  stack = [root]

  until stack.empty?
    root = stack.pop
    next unless root
    return true if equal_trees?(root, sub_root)
    stack << root.left << root.right
  end

  false
end

def equal_trees?(root1, root2)
  stack = [[root1, root2]]

  until stack.empty?
    root1, root2 = stack.pop
    next unless root1 || root2
    return false unless root1 && root2 && root1.val == root2.val
    stack << [root1.left, root2.left] << [root1.right, root2.right]
  end

  true
end

File: leetcode/subtree-of-another-tree/subtree-of-another-tree-naive-iter.cpp
// LeetCode #572 - Subtree of Another Tree
// https://leetcode.com/problems/subtree-of-another-tree/
// Naive solution, quadratic O(MN) runtime, implemented iteratively.

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    [[nodiscard]] static bool
    isSubtree(const TreeNode* root, const TreeNode* subRoot) noexcept;
};

namespace {
    [[nodiscard]] bool equal_trees(const TreeNode* root1,
                                   const TreeNode* root2) noexcept
    {
        auto stack = std::stack<tuple<const TreeNode*, const TreeNode*>>{};

        for (stack.emplace(root1, root2); !empty(stack); ) {
            tie(root1, root2) = stack.top();
            stack.pop();

            if (!root1 && !root2) continue;

            if (!root1 || !root2 || root1->val != root2->val) return false;

            stack.emplace(root1->left, root2->left);
            stack.emplace(root1->right, root2->right);
        }

        return true;
    }
}

bool Solution::isSubtree(const TreeNode* root,
                         const TreeNode* const subRoot) noexcept
{
    if (!subRoot) return !root;

    auto stack = std::stack<const TreeNode*>{};

    for (stack.push(root); !empty(stack); ) {
        root = stack.top();
        stack.pop();

        if (!root) continue;

        if (equal_trees(root, subRoot)) return true;

        stack.push(root->left);
        stack.push(root->right);
    }

    return false;
}
